activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb

```diff
@@ -77,7 +77,7 @@ def cache_sql(sql, binds)
         result =
           if @query_cache[sql].key?(binds)
             ActiveSupport::Notifications.instrument("sql.active_record",
-              :sql => sql, :binds => binds, :name => "CACHE", :connection_id => object_id)
+              :sql => sql.freeze, :binds => binds.freeze, :name => "CACHE", :connection_id => object_id)
```

Comment:
Can/should the caller do this freezing? Arguably, this is another manifestation of the same problem you were originally tracking: code mutating a value it received as a parameter, when it should really be leaving it alone.

The query cache lookups seem like they would benefit from using frozen values, too.